### PR TITLE
fix(react): bugs in field, checkbox, radio, textfield and dialog when running in React strict mode

### PR DESCRIPTION
**Pull request by unekinn**
## Summary

Since we found a bug (#3852) that only occurs in React strict mode, this PR
- Ensures our Storybook runs in strict mode
- Ensures our tests run in strict mode

The latter change led to six of our test suites failing. The failures can be divided into three categories:

**Category 1:** Components using `Field` gets duplicate ids i `aria-describedby`

```html
  <input
    aria-describedby=":md2ywi704b5:description:1 :md2ywi704b5:description:1"
                                              // ^^^^^^^^^^^^^^^^^^^^^^^^^^ THIS SHOULD NOT BE HERE
  />
```

This is fixed in `fieldObserver` by

```js
if (!describedbyIds.includes(el.id)) {
  // ...existing code to add id to describedbyIds array...
}
```

**Category 2:** Actual error i `dialog.tsx` hook cleanup

```tsx
    /* handle closing */
    useEffect(() => {
      const handleClose = (event: Event) => onClose?.(event);

      dialogRef.current?.addEventListener('close', handleClose);
      return () => dialogRef.current?.removeEventListener('close', handleClose);
    }, [onClose]);
```

What's wrong with this code? When the effect is performed twice due to StrictMode, `dialogRef.current` will be a different value on unmount than it is on mount. Thus cleanup isn't performed correctly, and `onClose` is called twice when close button or Esc is pressed.

Fix:

```tsx
    /* handle closing */
    useEffect(() => {
      const handleClose = (event: Event) => onClose?.(event);

      const currentRef = dialogRef.current;
      currentRef?.addEventListener('close', handleClose);
      return () => currentRef?.removeEventListener('close', handleClose);
    }, [onClose]);
```

**Category 3:** “Wrong” number of calls in `useEffect` init / cleanup

This happens in `useMediaQuery`, where we specifically test that the event listener is added once on mount, then removed on unmount.

E.g.

```ts
  it('Removes the event listener on unmount', () => {
    const removeEventListener = vi.fn();
    matchMediaValueMock({ removeEventListener });
    const { unmount } = renderHook(() => useMediaQuery(query));
    expect(removeEventListener).not.toHaveBeenCalled();
    unmount();
    expect(removeEventListener).toHaveBeenCalledTimes(1);
  });
```

We _could_ fix this by doing

```ts
expect(removeEventListener).toHaveBeenCalledTimes(1);
unmount();
expect(removeEventListener).toHaveBeenCalledTimes(2);
```

but that makes the test seem quite illogical. Since we are specifically testing mount / unmount I think it's better to disable strict mode here:

```ts
  it('Removes the event listener on unmount', () => {
    const removeEventListener = vi.fn();
    matchMediaValueMock({ removeEventListener });
    const { unmount } = renderHook(() => useMediaQuery(query), {
      // In React strict mode the hook is mounted twice. Disable it to properly test hook cleanup.
      reactStrictMode: false,
    });
    expect(removeEventListener).not.toHaveBeenCalled();
    unmount();
    expect(removeEventListener).toHaveBeenCalledTimes(1);
  });
```

## Checks

- [x] I have read the [contribution guidelines](https://github.com/digdir/designsystemet/blob/main/CONTRIBUTING.md)
- [x] I have added a changeset (run `pnpm changeset` if relevant)

### CODE REVIEW

**Comment by mimarz** [comment]
nice find! Releasing a snapshot of this so we can test downstream.

Released on `@test`. 

```
🦋  info Publishing "@digdir/designsystemet" at "0.0.0-test-20250714130823"
🦋  info Publishing "@digdir/designsystemet-css" at "0.0.0-test-20250714130823"
🦋  info Publishing "@digdir/designsystemet-react" at "0.0.0-test-20250714130823"
🦋  info Publishing "@digdir/designsystemet-theme" at "0.0.0-test-20250714130823"
```

**Review by mimarz — CHANGES_REQUESTED** [review]
> **Category 1:** Components using `Field` gets duplicate ids i `aria-describedby`
> 
> ```
>   <input
>     aria-describedby=":md2ywi704b5:description:1 :md2ywi704b5:description:1"
>                                               // ^^^^^^^^^^^^^^^^^^^^^^^^^^ THIS SHOULD NOT BE HERE
>   />
> ```
> 
> This is fixed in `fieldObserver` by
> 
> ```js
> if (!describedbyIds.includes(el.id)) {
>   // ...existing code to add id to describedbyIds array...
> }
> ```

This bug still seems to be there, but only when `data-field="validation"` is there.
Tested using our demo app with React 18.3.1. https://github.com/designsystemet/demo
image


> **Category 2:** Actual error i `dialog.tsx` hook cleanup

✅
> **Category 3:** “Wrong” number of calls in `useEffect` init / cleanup

✅

**Comment by unekinn** [comment]
> > **Category 1:** Components using `Field` gets duplicate ids i `aria-describedby`
> 
> This bug still seems to be there, but only when `data-field="validation"` is there.

Should be fixed now. We didn't have a test which had both description and error message, so I missed an edge case.

File: .changeset/cold-animals-clean.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"@digdir/designsystemet-react": patch
+---
+
+**Dialog**: avoid onClose being called twice when React is in strict mode
```
Comment (mimarz) [inline .changeset/cold-animals-clean.md]:
```suggestion
**Dialog**: Fix `onClose` being called twice when React is in strict mode
```

File: .changeset/fruity-spoons-thank.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+"@digdir/designsystemet-react": patch
+---
+
+**Field, Checkbox, Radio, Switch, Textfield**: avoid duplicate ids in aria-describedby when React is in strict mode
```
Comment (mimarz) [inline .changeset/fruity-spoons-thank.md]:
```suggestion
**Field, Checkbox, Radio, Switch, Textfield**: Fix duplicate ids in `aria-describedby` when React is in strict mode
```

**Review by mimarz — COMMENTED** [review]
Nice! Tested downstream again and no duplicate ids when using `Field` now! 

On a side not, should we update the last test in `field-observer.test.tsx` (L100) to omit the first matched ids from the `aria-describedby` string, then test that the string is "empty". Could be a good way for testing we only have one occurence of the id and not duplicates.

**Comment by unekinn** [comment]
> On a side not, should we update the last test in `field-observer.test.tsx` (L100) to omit the first matched ids from the `aria-describedby` string, then test that the string is "empty"? Could be a good way for testing we only have one occurence of the id and not duplicates.

@mimarz Good point that this should be tested in field-observer too. What do you think of my last commit now?

**Comment by mimarz** [comment]
> > On a side not, should we update the last test in `field-observer.test.tsx` (L100) to omit the first matched ids from the `aria-describedby` string, then test that the string is "empty"? Could be a good way for testing we only have one occurence of the id and not duplicates.
> 
> @mimarz Good point that this should be tested in field-observer too. What do you think of my last commit now?

nice! that worked even better!

**Review by mimarz — APPROVED** [review]
Looks good!
